tracks/julia/exercises/word-count/mentoring.md

```diff
@@ -121,4 +121,75 @@ function wordcount(sentence)
 end
 ```
 
+** Without Regex ** 
+
+This solution doesn't use a regex at all. It uses the default split behavior (which checks for `isspace` characters) and then strips punctuation characters left over on the "outside" of words. Filtering hoas to come after stripping in case the strip operation itself left an empty string.
+```julia
+function wordcount(sentence)
+    words = strip.(ispunct, split(lowercase(sentence)))
+    filter!(!isempty, words)
+    d = Dict{eltype(words), Int}()
+    for w in words
+        d[w] = get(d, w, 0) + 1
+    end
+    d
+end
+```
+It's possible to create the dictionary more efficiently with the `countmap` function from `StatsBase` (which only hashes the key once per check). With countmap, you can code-golf the above into:
+```julia
+using StatsBase: countmap
+
+wordcount_withstatsbase(sentence) = countmap(filter!(!isempty, strip.(ispunct, split(lowercase(sentence)))))
+```
+
+
+** Totally manual ** 
+
+This method iterates through the string one character at a time and switches behavior at word boundaries (when `iswordchar` changes from true to false or false to true). The dictionary has `SubString` keys in this version, which are like the string equivalent of Array `views`. It is about 20% faster than the StatsBase version on small inputs like those in the tests (performance is almost identicaly for inputs without thousands of characters). It's about 1.5x-2x faster than the next-fastest regex version, likely owing to the smaller number of allocations.
+```julia
+function wordcount(input)
+    input = lowercase(input)
+    d = Dict{SubString{typeof(input)}, Int}()
+    iswordchar(c) = !ispunct(c) && !isspace(c)
+
+    in_word = false
+    j = 0
+    for i in eachindex(input)
+        c = input[i]
+        # last-character is a special case, because there is no need
+        # for a lookahead and we need to record the last word.
+        if i == lastindex(input)
+            j = in_word ? j : i
+            i = iswordchar(c) ? i : i-1
+            if j <= i
+                k = SubString(input, j:i)
+                d[k] = get(d, k, 0) + 1
+            end
+            break
+        end
```

Comment:
I ran into an annoying edge case with final-character words. This takes care of that, but isn't too pretty.